retrieval/scrape.go

```diff
@@ -0,0 +1,292 @@
+	running sync.WaitGroup
+}
+
+func NewScrapePool(app storage.SampleAppender) *ScrapePool {
+	sp := &ScrapePool{
+		appender: app,
+		states:   map[model.Fingerprint]*TargetStatus{},
+		newLoop:  newScrapeLoop,
+	}
+	sp.ctx, sp.cancel = context.WithCancel(context.Background())
+
+	return sp
+}
+
+func (sp *ScrapePool) Stop() {
+	if sp.cancel != nil {
+		sp.cancel()
+	}
+	sp.running.Wait()
+}
+
+// Sync the running scrapers with the provided list of targets.
+func (sp *ScrapePool) Sync(targets []*Target) {
+	sp.mtx.Lock()
+	defer sp.mtx.Unlock()
+
+	// Shutdown all running loops after their current scrape
+	// is finished.
+	for _, l := range sp.loops {
+		go l.stop()
+	}
+
+	// Old loops are shutting down, create new from the targets.
+	sp.loops = map[model.Fingerprint]loop{}
+
+	for _, t := range targets {
+		fp := t.fingerprint()
+
+		// Create states for new targets.
+		state, ok := sp.states[fp]
+		if !ok {
+			state = &TargetStatus{}
+			sp.states[fp] = state
+		}
+
+		// Create and start loops the first time we see a target.
+		if _, ok := sp.loops[fp]; !ok {
+			loop := sp.newLoop(sp.appender, t, state)
+			sp.loops[fp] = loop
+
+			sp.running.Add(1)
+			go func(t *Target) {
+				loop.run(sp.ctx, t.interval(), t.timeout(), nil)
+				sp.running.Done()
+			}(t)
+		}
+	}
+
+	// Drop states of targets that no longer exist.
+	for fp := range sp.states {
+		if _, ok := sp.loops[fp]; !ok {
+			delete(sp.states, fp)
+		}
+	}
+}
+
+type loop interface {
+	run(ctx context.Context, interval, timeout time.Duration, errc chan<- error)
+	stop()
+}
+
+// scrapeLoop manages scraping cycles.
+type scrapeLoop struct {
+	scraper  scraper
+	state    *TargetStatus
+	appender storage.SampleAppender
+
+	// A function to calculate an initial offset for a given
+	// scrape interval.
+	offset func(time.Duration) time.Duration
+
+	stopc, done chan struct{}
+	cancel      func()
+	mtx         sync.RWMutex
+}
+
+func newScrapeLoop(app storage.SampleAppender, target *Target, state *TargetStatus) loop {
+	return &scrapeLoop{
+		scraper:  target,
+		appender: target.wrapAppender(app),
+		state:    state,
+		offset:   target.offset,
+	}
+}
+
+func (sl *scrapeLoop) active() bool {
+	return sl.cancel != nil
+}
+
+func (sl *scrapeLoop) run(ctx context.Context, interval, timeout time.Duration, errc chan<- error) {
+	if sl.active() {
+		return
+	}
+
+	// If no offset function is provided, default to starting immediately.
+	if sl.offset == nil {
+		sl.offset = func(time.Duration) time.Duration { return 0 }
+	}
+
+	ctx, sl.cancel = context.WithCancel(ctx)
+
+	sl.stopc = make(chan struct{})
+	sl.done = make(chan struct{})
+
+	defer func() {
+		close(sl.done)
+		sl.cancel = nil
+	}()
+
+	select {
+	case <-time.After(sl.offset(interval)):
+		// Continue after random scraping offset.
+	case <-ctx.Done():
+		return
+	case <-sl.stopc:
+		return
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		start := time.Now()
+
+		err := sl.scrape(ctx, timeout)
```

Comment:
If timeout > interval, the new logic could result in two scrapes in progress at once which won't end well. We should probably require that timeout <= interval.